server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/SpanServiceElasticsearch.java

```diff
@@ -0,0 +1,212 @@
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * @author Pavol Loffay
+ */
+public class SpanServiceElasticsearch implements SpanService {
+
+    private static final MsgLogger log = MsgLogger.LOGGER;
+
+    public static final String SPAN_TYPE = "span";
+
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    private ElasticsearchClient client = ElasticsearchClient.getSingleton();
+
+
+    @Override
+    public Span getSpan(String tenantId, String id) {
+
+        GetResponse response = null;
+        try {
+            response = client.getClient()
+                    .prepareGet(client.getIndex(tenantId), SPAN_TYPE, id)
+                    .setRouting(id)
+                    .execute()
+                    .actionGet();
+        } catch (IndexMissingException ex) {
+            log.errorf("Missing span index %s", tenantId);
+            return null;
+        }
+
+        Span span = null;
+        if (!response.isSourceEmpty()) {
+            try {
+                span = deserialize(response.getSourceAsString(), Span.class);
+            } catch (IOException ex) {
+                log.errorFailedToParse(ex);
+                throw new IllegalStateException("Failed to deserialize span", ex);
+            }
+        }
+
+        log.tracef("Get span with id[%s] is: %s", id, span);
+
+        return span;
+    }
+
+    @Override
+    public List<Span> getChildren(String tenantId, String id) {
+
+        List<Span> spans = new ArrayList<>();
+        final String index = client.getIndex(tenantId);
+        try {
+
+            RefreshRequestBuilder refreshRequestBuilder = client.getClient()
+                    .admin()
+                    .indices()
+                    .prepareRefresh(index);
+
+            client.getClient()
+                    .admin()
+                    .indices()
+                    .refresh(refreshRequestBuilder.request())
+                    .actionGet();
+
+            QueryBuilder query = QueryBuilders.termQuery("parentId", id);
+
+            SearchRequestBuilder request = client.getClient()
+                    .prepareSearch(index)
+                    .setTypes(SPAN_TYPE)
+                    .setSearchType(SearchType.DFS_QUERY_THEN_FETCH)
+                    .setQuery(query);
+
+            SearchResponse response = request.execute()
+                    .actionGet();
+
+            for (SearchHit searchHitFields : response.getHits()) {
+                try {
+                    spans.add(deserialize(searchHitFields.getSourceAsString(), Span.class));
+                } catch (IOException ex) {
+                    log.errorFailedToParse(ex);
+                    throw new IllegalStateException("Failed to deserialize span", ex);
+                }
+            }
+        } catch (IndexMissingException ex) {
+            log.errorf("No index[%s] found, so unable to retrieve spans", index);
```

Comment:
In general, methods that return a single object are meant to be called by someone who knows that this object should exist. So, any precondition has to be fulfilled, such as the index. If the index is not there, throw an exception, as it breaks your expectation (not the client's). If the record doesn't exist for that ID, you can choose between throwing an exception and returning null. I personally prefer to return null, as it allows for the client code to throw it if they wish. Exceptions are expensive, so, I'd avoid it for this case.

For a list of objects, it's a bit different: the calling code don't know if there are records on this collection, so, an empty collection means that the index exists, but there are no records there. Returning null means that the collection itself doesn't exist. Any precondition that is not fulfilled should render an exception. So, it's a matter of perspective: is the index part of the precondition, or is it the collection itself? If you see the index as being the collection, then returning null is appropriate if the collection/index doesn't exist. If you see the index as a precondition, then throw an exception.